Add support for specifying an entity's shortName

Nokel81 opened the pull request:
Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Add support for specifying an entity's shortName

- Remove CatalogEntityItem as it is a needless abstraction

- Refix transparent background bug after bad rebase

- Improve type safety around CatalogEntities by having CatalogEntitySpec
  be Partial<Record<string, unknown>>

- Cleanup implementation of catalogSyncToRendererInjectable

- Always use EntityPreferencesStore as an injectable

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Make CatalogEntityRegistry.init redundent
- Make all IPC injectable

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Ensure that catalog is initialized during startFrame

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Fix tests that overroad catalog sync
- Actually use the catalog sync

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Add behavioural tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Don't assign in a computed

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Make each store have a seperate migration version

- Allows each store to only bump its version when necessary
- Allows more decoupling from the extension API version so that PRs go stale slower

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Add behavioural tests for short name length

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Fix uses of testUsingTakeTime

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Fix test flakiness by removing animate delay in tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Update snapshots after removing timing sideeffects

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 commented on package.json:
```diff
@@ -3,7 +3,7 @@
   "description": "OpenLens - Open Source IDE for Kubernetes",
   "homepage": "https://github.com/lensapp/lens",
-  "version": "5.3.0",
+  "version": "5.4.0-alpha.1",
```
I need to do this so that I can add a new migration

Nokel81 commented on src/common/catalog/helpers.ts:
```diff
@@ -0,0 +1,71 @@
+  ].filter(Boolean).join("");
+}
+
+export function getShortName(entity: CatalogEntity): string {
```
I didn't really think that this should be a part of `CatalogEntity`

Nokel81 commented on src/main/catalog-pusher.ts:
```diff
@@ -25,15 +25,25 @@ import type { CatalogEntityRegistry } from "./catalog";
   broadcastMessage("catalog:items", items);
 }, 1_000, { leading: true, trailing: true });
 
 export function pushCatalogToRenderer(catalog: CatalogEntityRegistry) {
```
I made the modifications here so that we won't have to deal with issues like https://github.com/lensapp/lens/pull/4555.

That is because this acts as a pure reducer

Nokel81 commented on src/renderer/components/+catalog/catalog-entity-details.tsx:
```diff
@@ -27,32 +27,34 @@ import type { CatalogCategory, CatalogEntity } from "../../../common/catalog";
-interface Props<T extends CatalogEntity> {
-  item: CatalogEntityItem<T> | null | undefined;
+interface Props {
+  entity: CatalogEntity | null | undefined;
```
The main reason for removing `CatalogEntityItem` was because it was unnecessary. `CatalogItem` already implements the `ItemObject` interface so the second level of abstraction didn't gain us anything.

The only function that needed to be factored out was `getLabelBadges` but that can (and now is) easily a free function.

Nokel81 commented on src/renderer/components/+catalog/catalog-entity-details.tsx:
```diff
@@ -27,32 +27,34 @@ import type { CatalogCategory, CatalogEntity } from "../../../common/catalog";
-    const details = detailItems.map(({ components }, index) => {
-      return <components.Details entity={item.entity} key={index}/>;
-    });
+  renderContent(entity: CatalogEntity | undefined, onRun: () => void) {
```
The other change that was required by removing `CatalogEntityItem` was the injection of the `onRun` function here because it needs to go through the `catalogEntityRegistry` but so as to keep it injectable, I did not import that.

Nokel81 commented on src/renderer/components/avatar/avatar.tsx:
```diff
@@ -21,70 +21,20 @@
-import { cssNames, iter } from "../../utils";
+import { cssNames } from "../../utils";
 
 export interface AvatarProps extends HTMLAttributes<HTMLElement> {
```
These changes to `<Avatar />` were made to make it more reusable. Namely, I felt that there was too much business logic going on in here. Namely the computation of `shortName`'s.

Futhermore, because only `src` was a field (and not icon). There was a strange disconnect where the use of `<Avatar>` would include both passing in an optional `src` as well as optionally rendering an `<Icon />`. 

For that reason, I made `<EntityIcon>` to do all the work instead.

aleksfront commented on src/common/catalog/catalog-entity.ts:
```diff
@@ -38,14 +38,38 @@ export type CatalogEntityConstructor<Entity extends CatalogEntity> = (
+   * The specific version that the associated constructor is for. This MUST be
+   * a DNS label and SHOULD be of the form `vN`, `vNalphaY`, or `vNbetaY` where
+   * `N` and `Y` are both integers greater than 0.
+   */
```
This all relates to `name: string` prop? Maybe it needs some example.

Nokel81 replied:
Sure I can add some examples

aleksfront commented on src/common/catalog/catalog-entity.ts:
```diff
@@ -233,6 +265,14 @@ export abstract class CatalogEntity<
+
+  public isEnabled(): boolean {
+    return this.status.enabled ?? true;
+  }
```
If no `status.enabled` then `isEnabled` returns true? Shouldn't it be otherwise, just
```
return !!this.status.enabled
```

Nokel81 replied:
Nope, the documentation says that the default value is `true`

Nokel81 replied:
Plus this is what CatalogEnityitem already does

aleksfront commented on src/common/entity-preferences-store.ts:
```diff
@@ -0,0 +1,72 @@
+    if (!this.preferences.has(entityId)) {
+      this.preferences.set(entityId, preferences);
+    } else {
+      this.preferences.set(entityId, merge(this.preferences.get(entityId), preferences));
```
Should work fine without `if` checks above. With `undefined` value it will be merged properly. Try in repl https://lodash.com/docs/4.17.15#merge

aleksfront commented on src/extensions/registries/entity-setting-registry.ts:
```diff
@@ -32,39 +33,86 @@ export interface EntitySettingComponents {
+  /**
+   * If `true` then this item matches all versions, otherwise is a set of versions it supports
+   */
+  apiVersions: Set<string> | true;
```
These don't relate to using entity short name?

Nokel81 replied:
They do actually because I want to add the entity setting to all entity types.

aleksfront commented on src/common/catalog/helpers.ts:
```diff
@@ -0,0 +1,71 @@
+  return name.split(/@+/);
+}
+
+export function computeDefaultShortName(name: string) {
```
Since all methods are in separate helper file, it's a good opportunity to have unit test for them.

aleksfront reviewed (commented):
1. Can you add a bit more description of why we need specifying short names for entities?
2. Is it possible to separate removing `CatalogEntityItem` into separate PR?

jim-docker commented on src/common/hotbars/store.ts:
```diff
@@ -349,3 +374,23 @@ function ensureExactHotbarItemLength(hotbar: Hotbar) {
+ * @param hotbar The hotbar to modify
+ */
+function ensureNamesAndShortNames(hotbar: Hotbar) {
+  for (let i = 0; i < hotbar.items.length; i += 1) {
```
I can't remember the last time you used a for loop like this. no map, filter, etc.?

Nokel81 replied:
I guess I could have used `map`

Nokel81 replied:
Messes with the types too much if we have it as a map currently.

jim-docker commented on src/common/hotbars/types.ts:
```diff
@@ -7,15 +7,16 @@ import * as uuid from "uuid";
+export interface HotbarItemEntity {
+  uid: string;
+  name: string;
+  shortName: string;
```
breaking change?

Nokel81 replied:
Don't think this type is exported.

jim-docker replied:
Is `HotbarItem` exported?

Nokel81 replied:
No, and neither is `Hotbar`

jim-docker reviewed (commented):
This is massive, with changes added a year later. Can the recent commits be in a separate PR?

Shouldn't the short name be limited to a small number of characters, say 5?

This doesn't look good:
![Screen Shot 2023-01-18 at 2 56 09 PM](https://user-images.githubusercontent.com/40840436/213323756-1ad8a261-ec19-4474-a3a0-1ca044c795b0.png)

jim-docker commented on src/common/cluster-store/cluster-store.injectable.ts:
```diff
@@ -36,6 +37,7 @@ const clusterStoreInjectable = getInjectable({
     persistStateToConfig: di.inject(persistStateToConfigInjectionToken),
     enlistMessageChannelListener: di.inject(enlistMessageChannelListenerInjectionToken),
     shouldDisableSyncInListener: di.inject(shouldBaseStoreDisableSyncInIpcListenerInjectionToken),
+    sendMessageToChannel: di.inject(sendMessageToChannelInjectionToken),
```
There's obviously an existing pattern for this but it seems painful that this change has to be done for every store.

Nokel81 replied:
Yes I agree this is quite painfull.

I think a better solution would be for `fromStore` and `toJson` to be configuration params to `BaseStore` instead of `abstract` methods.

That way we could have one `createStoreInjectable` that is referenced instead of all of the dependencies being mandatory in many places.

Namely, composition over inheritance.

jim-docker commented on src/extensions/extensions-store/extensions-store.injectable.ts:
```diff
@@ -27,6 +28,7 @@ const extensionsStoreInjectable = getInjectable({
     persistStateToConfig: di.inject(persistStateToConfigInjectionToken),
     enlistMessageChannelListener: di.inject(enlistMessageChannelListenerInjectionToken),
     shouldDisableSyncInListener: di.inject(shouldBaseStoreDisableSyncInIpcListenerInjectionToken),
+    sendMessageToChannel: di.inject(sendMessageToChannelInjectionToken),
```
Isn't this already done in the `ExtensionStore` constructor?

Nokel81 replied:
Haha different type all together. I get messed up too.

There is `ExtensionStore` which is exported to the extension API and then there is `ExtensionsStore` which keeps track of if an extension is enabled or not.

jim-docker replied:
oops

jim-docker reviewed (commented):
Bigger still. Maybe @jansav or @Iku-turso should have a quick look too?

Nokel81 committed:
imerge 'issue-3498': manual merge 1-1

# Conflicts:
#	package.json
#	packages/core/src/common/catalog/__tests__/helpers.test.ts
#	packages/core/src/common/catalog/helpers.ts
#	packages/core/src/common/cluster-store/migration-version.injectable.ts
#	packages/core/src/common/hotbars/migration-version.injectable.ts
#	packages/core/src/common/ipc/catalog.ts
#	packages/core/src/common/user-store/migration-version.injectable.ts
#	packages/core/src/common/vars/store-migration-version.injectable.ts
#	packages/core/src/common/weblinks-store/migration-version.injectable.ts
#	packages/core/src/extensions/extension-loader/file-system-provisioner-store/migration-version.injectable.ts
#	packages/core/src/extensions/extensions-store/migration-version.injectable.ts
#	packages/core/src/features/entity-settings/__snapshots__/limit-short-name-length.test.tsx.snap
#	packages/core/src/features/entity-settings/limit-short-name-length.test.tsx
#	packages/core/src/features/hotbar/__snapshots__/shows-custom-entity-names.test.tsx.snap
#	packages/core/src/features/hotbar/shows-custom-entity-names.test.tsx
#	packages/core/src/main/catalog-sync-to-renderer/broadcaster.global-override-for-injectable.ts
#	packages/core/src/main/catalog-sync-to-renderer/broadcaster.injectable.ts
#	packages/core/src/main/catalog-sync-to-renderer/push-catalog-to-renderer.injectable.ts
#	packages/core/src/main/catalog/index.ts
#	packages/core/src/renderer/components/+entity-settings/internal-kubernetes-cluster/general-settings.injectable.tsx
#	packages/core/src/renderer/components/+entity-settings/internal-kubernetes-cluster/metrics-settings.injectable.tsx
#	packages/core/src/renderer/components/+entity-settings/internal-kubernetes-cluster/namespace-settings.injectable.tsx
#	packages/core/src/renderer/components/+entity-settings/internal-kubernetes-cluster/node-shell-settings.injectable.tsx
#	packages/core/src/renderer/components/+entity-settings/internal-kubernetes-cluster/proxy-settings.injectable.tsx
#	packages/core/src/renderer/components/+entity-settings/internal-kubernetes-cluster/terminal-settings.injectable.tsx
#	packages/core/src/renderer/components/animate/is-enabled.global-override-for-injectable.ts
#	packages/core/src/renderer/components/animate/is-enabled.injectable.ts
#	packages/core/src/renderer/components/avatar/compute-random-color.global-override-for-injectable.ts
#	packages/core/src/renderer/components/avatar/compute-random-color.injectable.ts
#	src/main/catalog/index.ts
#	src/renderer/components/+entity-settings/internal-kubernetes-cluster/general-settings.injectable.tsx
#	src/renderer/components/+entity-settings/internal-kubernetes-cluster/metrics-settings.injectable.tsx
#	src/renderer/components/+entity-settings/internal-kubernetes-cluster/namespace-settings.injectable.tsx
#	src/renderer/components/+entity-settings/internal-kubernetes-cluster/node-shell-settings.injectable.tsx
#	src/renderer/components/+entity-settings/internal-kubernetes-cluster/proxy-settings.injectable.tsx
#	src/renderer/components/+entity-settings/internal-kubernetes-cluster/terminal-settings.injectable.tsx
#	src/renderer/components/+entity-settings/internal/kubernetes-cluster/general-settings.injectable.tsx
#	src/renderer/components/+entity-settings/internal/kubernetes-cluster/metrics-settings.injectable.tsx
#	src/renderer/components/+entity-settings/internal/kubernetes-cluster/namespace-settings.injectable.tsx
#	src/renderer/components/+entity-settings/internal/kubernetes-cluster/node-shell-settings.injectable.tsx
#	src/renderer/components/+entity-settings/internal/kubernetes-cluster/proxy-settings.injectable.tsx
#	src/renderer/components/+entity-settings/internal/kubernetes-cluster/terminal-settings.injectable.tsx
#	src/renderer/components/entity-icon/index.ts
#	yarn.lock

Nokel81 committed:
Readd dep bump

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Merge branch 'master' into issue-3498

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Merge branch 'master' into issue-3498

Nokel81 committed:
Fix package deps

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Merge branch 'master' into issue-3498

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Merge branch 'master' into issue-3498

Nokel81 committed:
Merge branch 'master' into issue-3498

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Merge remote-tracking branch 'origin/master' into issue-3498

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Update lock file

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Merge remote-tracking branch 'origin/master' into issue-3498

Signed-off-by: Sebastian Malton <sebastian@malton.name>
